ssl/statem/extensions_clnt.c

```diff
@@ -817,7 +817,7 @@ EXT_RETURN tls_construct_ctos_early_data(SSL *s, WPACKET *pkt,
     s->psksession = psksess;
     if (psksess != NULL) {
         OPENSSL_free(s->psksession_id);
-        s->psksession_id = OPENSSL_memdup(id, idlen);
+        s->psksession_id = OPENSSL_memdup(id, idlen + 1);
```

Comment:
This doesn't look like the right fix to me. If we're using the new callbacks then the identity is not NULL terminated and we just read past the end of the user supplied buffer by 1 byte. It does not seem necessary to duplicate the NUL terminator because `s->psksession_id` is not expected to be NUL terminated at any point.

The root of the problem here is that TLSv1.2 allows an identity to be zero length. However TLSv1.3 requires an identity to be at least 1 byte. I don't see a way that we can seemlessly allow the old style callbacks work in TLSv1.3 if they return an empty identity. The question is what to do about it?